File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    var answer = 0;
    // target이 words 내부에 없는 경우는 0 리턴
    console.log()
    if (!words.includes(target)) {
        return 0
    }
    
    return bfs(begin, target, words);
}

const bfs = (begin, target, words) => {
    // 시작 단어부터 queue에 넣고, step은 0으로 초기화한다.
    queue = [];
    queue.push([begin, 0]) // 시작 단어, 단계
    
    while (queue.length > 0) {
        [now, step] = queue.shift(); // 현재 단어, 단계 꺼내기
        
        if (now === target) { // 타겟과 같다면 마무리
            return step;
        }
        
        // 단어를 모두 체크하면서, 해당 단어가 변경될 수 있는지 확인한다.
        words.forEach((word) => {
            let count = 0; // 같지 않은 부분 개수
            for (let i = 0; i < now.length ; i++) {
                // 단어에 알파벳 한개씩 체크해서, 같지 않은 부분 카운팅
                if (now[i] !== word[i]) {
                    count++;
                }
            }
            
            // 바꿀 수 있는 단어가 나온다면, 큐에 값 추가
            if (count === 1) {
                queue.push([word, step+1])
            }
        })
        
    }
    
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    answer = 0
    
    # 1. 변환할 수 없는 경우는 0 return
    # -> 즉, words 안에 target이 없다면 0 리턴
    if target not in words:
        return 0    
    # bfs를 이용해서 step을 기록해서 
    # 1개만 다른 단어를 찾으면 queue에 넣고 변환과정을 넣는다
    # 이게참… 신기하단말이지 무한반복될줄 알았는데
    return bfs(begin, target, words)

def bfs(begin, target, words):
    queue = deque()
    queue.append([begin, 0]) # 큐 선언
    
    
    # 큐에 변환 가능한 단어가 있을때까지 무한반복
    while queue:
        # print(queue)
        # 단어와 현재 단계 저장
        now, step = queue.popleft() # 값 저장 (현재 단어, step)
        
        # 만약 현재 찾은 단어가 목표 단어라면, 반복 마무리 후 리턴
        if now == target:
            return step

        
        # words 배열을 순회하면서 단어 바꿔가면서 진행하기 
        for word in words:
            count = 0 # 같지 않은 단어 자리 개수 
            for i in range(len(word)):
                # 단어가 1개만 다르다면 그 단어 추가 
                if now[i] != word[i]:
                    count += 1
            
            # 만약 같지 않은 단어가 한개라면, (즉 나머지가 모두 일치하다면, 한 자리 단어만 바꿀수있음)
            if count == 1:
                # 큐에 바꾸게 될 값 추가
                queue.append([word, step+1]) # 그리고 다음 변환 단계로 진행
            
       
    return 0
    